File: hooks/installer.py
import logging
import shutil
import time
from os import environ
import uuid
from os import path
from os.path import isdir, join
from subprocess import check_output

import requests_unixsocket
from bs4 import BeautifulSoup
from syncloudlib import fs, linux, gen, logger
from syncloudlib.application import paths, urls, storage, users

from database import Database

APP_NAME = 'mastodon'
USER_NAME = APP_NAME
PSQL_PATH = 'postgresql/bin/psql.sh'
PSQL_PORT = 5434
DB_USER = APP_NAME
DB_PASS = APP_NAME
DB_NAME = APP_NAME

logger.init(logging.DEBUG, console=True, line_format='%(message)s')



class Installer:
    def __init__(self):
        self.app_dir = paths.get_app_dir(APP_NAME)
        self.common_dir = paths.get_data_dir(APP_NAME)
        self.data_dir = join('/var/snap', APP_NAME, 'current')
        self.log = logger.get_logger(APP_NAME)
        self.config_dir = join(self.data_dir, 'config')
        self.db = Database(self.app_dir, self.data_dir, self.config_dir, join(self.app_dir, PSQL_PATH), DB_USER, PSQL_PORT)
        self.mastodon_dir = join(self.app_dir, 'ruby', 'mastodon')
        self.rails = join(self.mastodon_dir, 'bin', 'rails')
        self.install_file = join(self.common_dir, 'installed')
        environ['RAILS_ENV'] = 'production'

    def init_config(self):
        linux.useradd(USER_NAME)

        app_config_dir = join(self.app_dir, 'config')
        fs.makepath(join(self.data_dir, 'redis'))
        fs.makepath(join(self.data_dir, 'nginx'))
        fs.makepath(join(self.data_dir, 'nginx', 'cache'))

        storage_dir = storage.init_storage(APP_NAME, USER_NAME)
        variables = {
            'app_dir': self.app_dir,
            'common_dir': self.common_dir,
            'database_dir': self.db.database_dir,
            'db_psql_port': PSQL_PORT,
            'db_name': DB_NAME,
            'db_user': DB_USER,
            'db_password': DB_PASS,
            'storage_dir': storage_dir,
            'data_dir': self.data_dir,
            'app_url': urls.get_app_url(APP_NAME),
            'app_domain': urls.get_app_domain_name(APP_NAME),
            'secret': uuid.uuid4().hex,
            'secret_base': uuid.uuid4().hex,
            'disable_registration': False
        }
        gen.generate_files(app_config_dir, self.config_dir, variables)
        fs.chownpath(self.common_dir, USER_NAME, recursive=True)
        fs.chownpath(self.data_dir, USER_NAME, recursive=True)

    def install(self):
        self.log_common('before.install')
        self.init_config()
        self.db.init()
        self.db.init_config()
        self.log_common('after.install')

    def pre_refresh(self):
        self.db.backup()

    def post_refresh(self):
        self.log.info('post refresh')
        self.init_config()
        self.db.remove()
        self.db.init()
        self.db.init_config()

    def installed(self):
        return path.isfile(self.install_file)

    def configure(self):
        self.log_common('before.configure')
        self.log.info('configure')
        if self.installed():
            self.upgrade()
        else:
            self.initialize()
        self.log_common('after.configure')

    def upgrade(self):
        self.log.info('upgrade')
        self.db.restore()
        check_output([self.rails, 'db:migrate'], cwd=self.mastodon_dir)
        self.update_db_version()

    def initialize(self):
        self.log.info('initialize')

        self.db.execute('postgres', "ALTER USER {0} WITH PASSWORD '{1}';".format(DB_USER, DB_PASS))
        check_output([self.rails, 'db:setup'], cwd=self.mastodon_dir)
        check_output([self.rails, 'db:migrate'], cwd=self.mastodon_dir)
        self.update_db_version()
        with open(self.install_file, 'w') as f:
            f.write('installed\n')

    def update_db_version(self):
        shutil.copy(join(self.app_dir, 'version'), self.data_dir)

    def prepare_storage(self):
        storage.init_storage(APP_NAME, USER_NAME)

    def on_domain_change(self):
        self.log.info('domain change')
        self.init_config()

    def backup_pre_stop(self):
        self.pre_refresh()

    def restore_pre_start(self):
        self.post_refresh()

    def restore_post_start(self):
        self.configure()

    def log_common(self, name):
        with open(join(self.common_dir, '{0}.log'.format(name)), 'w') as f:
            f.write(str(check_output('ls -la', cwd=self.common_dir, shell=True)))

File: cli/installer/installer.go
package installer

import (
	"fmt"
	"github.com/google/uuid"
	cp "github.com/otiai10/copy"
	"github.com/syncloud/golib/config"
	"github.com/syncloud/golib/linux"
	"github.com/syncloud/golib/platform"
	"go.uber.org/zap"
	"os"
	"path"
)

const (
	App      = "mastodon"
	PsqlPort = 5434
)

type Variables struct {
	AppDir                                  string
	CommonDir                               string
	DatabaseDir                             string
	DbPsqlPort                              int
	DbName                                  string
	DbUser                                  string
	DbPassword                              string
	StorageDir                              string
	DataDir                                 string
	AppUrl                                  string
	AppDomain                               string
	Secret                                  string
	SecretBase                              string
	ActiveRecordEncryptionDeterministicKey  string
	ActiveRecordEncryptionKeyDerivationSalt string
	ActiveRecordEncryptionPrimaryKey        string
}

type Installer struct {
	newVersionFile     string
	currentVersionFile string
	configDir          string
	appDir             string
	dataDir            string
	commonDir          string
	mastodonDir        string
	railsBin           string
	platformClient     *platform.Client
	database           *Database
	installFile        string
	executor           *Executor
	logger             *zap.Logger
}

func New(logger *zap.Logger) *Installer {
	appDir := fmt.Sprintf("/snap/%s/current", App)
	dataDir := fmt.Sprintf("/var/snap/%s/current", App)
	commonDir := fmt.Sprintf("/var/snap/%s/common", App)
	mastodonDir := path.Join(appDir, "ruby", "mastodon")
	railsBin := path.Join(mastodonDir, "bin", "rails")

	configDir := path.Join(dataDir, "config")

	executor := NewExecutor(logger)
	return &Installer{
		newVersionFile:     path.Join(appDir, "version"),
		currentVersionFile: path.Join(dataDir, "version"),
		configDir:          configDir,
		appDir:             appDir,
		dataDir:            dataDir,
		commonDir:          commonDir,
		mastodonDir:        mastodonDir,
		railsBin:           railsBin,
		platformClient:     platform.New(),
		database:           NewDatabase(appDir, dataDir, configDir, App, executor, logger),
		installFile:        path.Join(commonDir, "installed"),
		executor:           executor,
		logger:             logger,
	}
}

func (i *Installer) Install() error {

	err := i.UpdateConfigs()
	if err != nil {
		return err
	}

	err = i.database.Init()
	if err != nil {
		return err
	}
	err = i.database.InitConfig()
	if err != nil {
		return err
	}

	return nil
}

func (i *Installer) Configure() error {
	if i.IsInstalled() {
		err := i.Upgrade()
		if err != nil {
			return err
		}
	} else {
		err := i.Initialize()
		if err != nil {
			return err
		}
	}

	return nil
}

func (i *Installer) IsInstalled() bool {
	_, err := os.Stat(i.installFile)
	return err == nil
}

func (i *Installer) Initialize() error {
	_, err := i.StorageChange()
	if err != nil {
		return err
	}

	err = i.database.Execute(
		"postgres",
		fmt.Sprintf("ALTER USER %s WITH PASSWORD '%s'", App, App),
	)
	if err != nil {
		return err
	}

	//err = i.database.createDbIfMissing(App)
	//if err != nil {
	//	return err
	//}

	err = i.executor.Run(i.railsBin, i.mastodonDir, "db:setup")
	if err != nil {
		return err
	}
	err = i.executor.Run(i.railsBin, i.mastodonDir, "db:migrate")
	if err != nil {
		return err
	}

	//err = i.database.Execute("postgres", fmt.Sprintf("GRANT CREATE ON SCHEMA public TO %s", App))
	//if err != nil {
	//	return err
	//}

	err = os.WriteFile(i.installFile, []byte("installed"), 0644)
	if err != nil {
		return err
	}

	return i.UpdateVersion()
}

func (i *Installer) Upgrade() error {
	err := i.database.Restore()
	if err != nil {
		return err
	}
	_, err = i.StorageChange()
	if err != nil {
		return err
	}
	//err = i.database.createDbIfMissing(App)
	//if err != nil {
	//	return err
	//}
	err = i.executor.Run(i.railsBin, i.mastodonDir, "db:migrate")
	if err != nil {
		return err
	}
	return i.UpdateVersion()
}

func (i *Installer) PreRefresh() error {
	return i.database.Backup()
}

func (i *Installer) PostRefresh() error {
	err := i.UpdateConfigs()
	if err != nil {
		return err
	}
	err = i.database.Remove()
	if err != nil {
		return err
	}
	err = i.database.Init()
	if err != nil {
		return err
	}
	err = i.database.InitConfig()
	if err != nil {
		return err
	}

	err = i.ClearVersion()
	if err != nil {
		return err
	}

	err = i.FixPermissions()
	if err != nil {
		return err
	}
	return nil

}
func (i *Installer) AccessChange() error {
	err := i.UpdateConfigs()
	if err != nil {
		return err
	}
	return err
}

func (i *Installer) StorageChange() (string, error) {
	return i.platformClient.InitStorage(App, App)
}

func (i *Installer) ClearVersion() error {
	return os.RemoveAll(i.currentVersionFile)
}

func (i *Installer) UpdateVersion() error {
	return cp.Copy(i.newVersionFile, i.currentVersionFile)
}

func (i *Installer) UpdateConfigs() error {
	err := linux.CreateUser(App)
	if err != nil {
		return err
	}

	storageDir, err := i.StorageChange()
	if err != nil {
		return err
	}

	err = linux.CreateMissingDirs(
		path.Join(i.dataDir, "nginx/cache"),
		path.Join(i.dataDir, "redis"),
		path.Join(i.dataDir, "system"),
	)
	if err != nil {
		return err
	}

	url, err := i.platformClient.GetAppUrl(App)
	if err != nil {
		return err
	}

	domain, err := i.platformClient.GetAppDomainName(App)
	if err != nil {
		return err
	}

	secret, err := getOrCreateUuid(path.Join(i.dataDir, ".secret"))
	if err != nil {
		return err
	}

	secretBase, err := getOrCreateUuid(path.Join(i.dataDir, ".secret.base"))
	if err != nil {
		return err
	}

	activeRecordEncryptionDeterministicKey, err := getOrCreateUuid(path.Join(i.dataDir, ".activeRecordEncryptionDeterministicKey"))
	if err != nil {
		return err
	}
	activeRecordEncryptionKeyDerivationSalt, err := getOrCreateUuid(path.Join(i.dataDir, ".activeRecordEncryptionKeyDerivationSalt"))
	if err != nil {
		return err
	}
	activeRecordEncryptionPrimaryKey, err := getOrCreateUuid(path.Join(i.dataDir, ".activeRecordEncryptionPrimaryKey"))
	if err != nil {
		return err
	}

	variables := Variables{
		AppDir:                                  i.appDir,
		CommonDir:                               i.commonDir,
		DatabaseDir:                             i.database.DatabaseDir(),
		DbPsqlPort:                              PsqlPort,
		DbName:                                  App,
		DbUser:                                  App,
		DbPassword:                              App,
		StorageDir:                              storageDir,
		DataDir:                                 i.dataDir,
		AppUrl:                                  url,
		AppDomain:                               domain,
		Secret:                                  secret,
		SecretBase:                              secretBase,
		ActiveRecordEncryptionDeterministicKey:  activeRecordEncryptionDeterministicKey,
		ActiveRecordEncryptionKeyDerivationSalt: activeRecordEncryptionKeyDerivationSalt,
		ActiveRecordEncryptionPrimaryKey:        activeRecordEncryptionPrimaryKey,
	}

	err = config.Generate(
		path.Join(i.appDir, "config"),
		path.Join(i.dataDir, "config"),
		variables,
	)
	if err != nil {
		return err
	}

	err = i.FixPermissions()
	if err != nil {
		return err
	}

	return nil

}

func (i *Installer) FixPermissions() error {
	err := linux.Chown(i.dataDir, App)
	if err != nil {
		return err
	}
	err = linux.Chown(i.commonDir, App)
	if err != nil {
		return err
	}
	return nil
}

func (i *Installer) BackupPreStop() error {
	return i.PreRefresh()
}

func (i *Installer) RestorePreStart() error {
	return i.PostRefresh()
}

func (i *Installer) RestorePostStart() error {
	return i.Configure()
}

func getOrCreateUuid(file string) (string, error) {
	_, err := os.Stat(file)
	if os.IsNotExist(err) {
		secret := uuid.New().String()
		err = os.WriteFile(file, []byte(secret), 0644)
		return secret, err
	}
	content, err := os.ReadFile(file)
	if err != nil {
		return "", err
	}
	return string(content), nil
}
